evm/vm/base.py

```diff
@@ -133,45 +133,48 @@ def execute_bytecode(self,
     #
     def import_block(self, block):
         self.receipts = []
-        self.configure_header(
+        self.block = self.block.copy(header=self.configure_header(
             coinbase=block.header.coinbase,
             gas_limit=block.header.gas_limit,
             timestamp=block.header.timestamp,
             extra_data=block.header.extra_data,
             mix_hash=block.header.mix_hash,
             nonce=block.header.nonce,
             uncles_hash=keccak(rlp.encode(block.uncles)),
-        )
+        ))
 
         # run all of the transactions.
         for transaction in block.transactions:
             self.apply_transaction(transaction)
 
         # transfer the list of uncles.
-        self.block.uncles = block.uncles
+        self.block = self.block.copy(
+            uncles=block.uncles,
+            transactions=block.transactions,
+            header=self.block.header,
+        )
```

Comment:
This repeated assignment of `self.block` suggests to me that this might be another place for the context manager.

```py
        block_builder = self.block.build_copy(
              transactions=block.transactions,
              uncles=block.uncles,
        )
        with block_builder as next_block:
            next_block.header = self.configure_header(**block.header.as_dict())
 
            # run all of the transactions.
            for transaction in next_block.transactions:
                self.apply_transaction(transaction, to_block=next_block)

         self.block = next_block
```

So either we:
1. embrace the mutable block within this context, and have `apply_transaction()` modify the block in place (shown above, but still not assign to `self.block`), or...
2. return a copy of the new block after applying the transaction